lang/funcs/crypto.go

```diff
@@ -155,24 +154,18 @@ var RsaDecryptFunc = function.New(&function.Spec{
 			return cty.UnknownVal(cty.String), fmt.Errorf("failed to decode input %q: cipher text must be base64-encoded", s)
 		}
 
-		block, _ := pem.Decode([]byte(key))
-		if block == nil {
-			return cty.UnknownVal(cty.String), fmt.Errorf("failed to parse key: no key found")
-		}
-		if block.Headers["Proc-Type"] == "4,ENCRYPTED" {
-			return cty.UnknownVal(cty.String), fmt.Errorf(
-				"failed to parse key: password protected keys are not supported. Please decrypt the key prior to use",
-			)
-		}
-
-		x509Key, err := x509.ParsePKCS1PrivateKey(block.Bytes)
+		rawKey, err := ssh.ParseRawPrivateKey([]byte(key))
 		if err != nil {
-			return cty.UnknownVal(cty.String), err
+			return cty.UnknownVal(cty.String), fmt.Errorf("failed to parse key: %s", err)
```

Comment:
is the Errorf text needed here? Just wondering if we're going to get errors like `failed to parse key: fail to parse SSH key`

Reply:
I had kinda hoped that the `encoding/asn1` package would produce machine-readable error values that we could use to present something more appropriate for the Terraform end-user audience, 'cause from what I see in #24970 these messages seem pretty poor, but sadly it looks like [`StructuralError`](https://pkg.go.dev/encoding/asn1?tab=doc#SyntaxError) and [`SyntaxError`](https://pkg.go.dev/encoding/asn1?tab=doc#SyntaxError) are both just wrappers around opaque strings formatted in the internals of the parser. :confounded: 

With that said, I reviewed [the parser code](https://github.com/golang/go/blob/go1.14.4/src/encoding/asn1/asn1.go#L1082) and it looks like this "tags don't match" error is a particularly ugly example but most of them are reasonably messages that suffer only from a total lack of useful context, like "base 128 integer too large". So the two forms this should end up producing are:

* `failed to parse key: asn1: structure error: tags don't match
(16 vs {class:1 tag:15 length:112 isCompound:true}) {optional:false
explicit:false application:false private:false defaultValue:<nil> tag:<nil>
stringType:0 timeType:0 set:false omitEmpty:false} pkcs1PrivateKey @2.`
* `failed to parse key: asn1: syntax error: invalid boolean`

I think the only way we could marginally improve these with the information we have is to recognize when the error type is either `asn1.SyntaxError` or asn1.StructuralError` and generate a different prefix (replacing `asn1: syntax error:`) that is more inline with our usual writing style, like:

* `invalid ASN1 data in the given private key: invalid boolean`
* `invalid ASN1 data in the given private key: tags don't match
(16 vs {class:1 tag:15 length:112 isCompound:true}) {optional:false
explicit:false application:false private:false defaultValue:<nil> tag:<nil>
stringType:0 timeType:0 set:false omitEmpty:false} pkcs1PrivateKey @2.`

(the second one isn't going to be nice _whatever_ we do to it, of course...)

A further embellishment, though very much _not necessary_, would be to use `function.NewArgErrorf` to associate this error explicitly with the second argument (argument index 1), and then HCL should indicate that argument's experession in particular as being the problematic part of the configuration.

Reply:
Thanks for the suggestions!

I think it is important to process these errors rather than just pass them through, because the crypto errors are pretty obscure. Adding even basic error annotations was the first step to debugging this.

I've made the suggested changes and I think it's a big improvement:

```shellsession
velara:24970 $ tf apply -var-file=trunc.tfvars

Error: Invalid function argument

  on example.tf line 13, in output "password_decrypted":
  13:   value=rsadecrypt(var.password_data, file("${path.module}/${var.rsa_private_key_file}") )
    |----------------
    | path.module is "."
    | var.rsa_private_key_file is "id_rsa_trunc"

Invalid value for "privatekey" parameter: invalid ASN1 data in the given
private key: data truncated.
```